### ISSUE TITLE
Deprecate and remove the `Scoped` type class

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#5131]
- [ ] Systematically remove `Scoped`.
- [ ] Systematically remove `'Static` (and with it the `Region` expression).

**Comment by paulbutcher** [comment]
@magnus-madsen Can I confirm that my understanding is correct before pushing ahead with this.

Take, for example, `Array.scanLeft`, which (after @stephentetley's changes in #5143) looks like this:

```
    pub def scanLeft(f: (b, a) -> b \ ef, s: b, arr: Array[a, r]): Array[b, r] \ { ef, Read(r), Write(r) } =
        let len = length(arr) + 1;
        let r = Scoped.regionOf(arr);
        let b = new(r, s, len);
        def loop(i, acc) = {
            if (i >= len)
                ()
            else {
                let s1 = f(acc, arr[i - 1]);
                b[i] = s1;
                loop(i + 1, s1)
            }
        };
        loop(1, s);
        b
```

I think that the right solution is to add an additional region parameter to this function (and most other functions that still use `regionOf`). Do I understand correctly?

**Comment by magnus-madsen** [comment]
> I think that the right solution is to add an additional region parameter to this function (and most other functions that still use `regionOf`). Do I understand correctly?

Yes, I think so.

**Comment by paulbutcher** [comment]
Thanks for confirming.

This leads to a second question about consistency. Take `MutList`. We will need to add a region parameter to [`MutList.insert!`](https://github.com/flix/flix/blob/master/main/src/library/MutList.flix#L575) because it currently uses `regionOf`. But there are other similar functions within `MutList` which don't (e.g. [`MutList.pop!`](https://github.com/flix/flix/blob/master/main/src/library/MutList.flix#L561).

Should we add a region parameter to all these functions for consistency? Or only where it's strictly necessary?

**Comment by paulbutcher** [comment]
Hmm. We may be premature with this. One of the functions that will need changing is [`Iterable.iterator`](https://github.com/flix/flix/blob/master/main/src/library/Iterable.flix#L30-L31), and doing that interacts with the code generated for `foreach`.

Given that we're planning to re-do `foreach` as part of #5132, perhaps we should wait until that's out of the way before tackling this?

**Comment by paulbutcher** [comment]
(or, alternatively, we could do everything else, leaving `Iterable` as-is)?

**Comment by magnus-madsen** [comment]
I think some data structures- including `MutList` will need to keep their region inside them. 

This is the only way they will be able to re-allocate some internal array.

**Comment by magnus-madsen** [comment]
In other words, in some cases, use of Scoped should be replaced by storing the capability in the structure and then reusing that instead of adding a new region parameter.

**Comment by magnus-madsen** [comment]
`Iterable.iterator` will defininitely need to take a region argument and think we can already make that change.

**Comment by paulbutcher** [comment]
> Iterable.iterator will defininitely need to take a region argument and think we can already make that change.

That will require changes to the code generated for `foreach`. Do you think that that's worth doing given that #5132 is on the horizon?

**Comment by paulbutcher** [comment]
> In other words, in some cases, use of Scoped should be replaced by storing the capability in the structure and then reusing that instead of adding a new region parameter.

Can you help me understand which cases are which?

**Comment by magnus-madsen** [comment]
> > In other words, in some cases, use of Scoped should be replaced by storing the capability in the structure and then reusing that instead of adding a new region parameter.
> 
> Can you help me understand which cases are which?

Yes. Think of it this way:

- Returns a new mutable data structure (e.g. a new array or MutList) => Needs region parameter of where to allocate.
- Modifies some internal state (e.g. MutList.push!) => Does not need region parameter. Should use an internally stored on.

**Comment by magnus-madsen** [comment]
> > Iterable.iterator will defininitely need to take a region argument and think we can already make that change.
> 
> That will require changes to the code generated for `foreach`. Do you think that that's worth doing given that #5132 is on the horizon?

Depends. No-one is assigned to that ticket yet :)

In any case, I think the refactoring might be useful and revealing for when we eventually have to implement `Stream`.

**Comment by stephentetley** [comment]
@paulbutcher - that definition of `scanLeft` is wrong, I must have missed changing it during the refactor. Because `scanLeft` creates a new array for the answer it should be supplied with a region to create it in.

Iterators and `Iterable.iterator` were deliberately left as-is because they caused problems with `foreach` comprehensions.

**Comment by paulbutcher** [comment]
> Yes. Think of it this way:
> 
> * Returns a new mutable data structure (e.g. a new array or MutList) => Needs region parameter of where to allocate.
> * Modifies some internal state (e.g. MutList.push!) => Does not need region parameter. Should use an internally stored on.

That makes sense. Thanks.

**Comment by paulbutcher** [comment]
Cool. So I suggest that I push ahead with this, leaving `Iterable.iterator` for a separate change.

**Comment by paulbutcher** [comment]
@magnus-madsen a large number of our tests currently return a region-carrying type, for example:

https://github.com/flix/flix/blob/master/main/test/ca/uwaterloo/flix/library/TestStringBuilder.flix#L23
https://github.com/flix/flix/blob/master/main/test/flix/Test.Exp.Null.flix#L13
https://github.com/flix/flix/blob/master/main/test/flix/Test.Exp.New.flix#L13

It's not clear to me how to modify these tests so that they don't use `Static`. Any ideas?

**Comment by magnus-madsen** [comment]
They have to be rewritten a bit. Take for example:

```
    @test
    def testNewMutList1Region03(): MutList[String, Static] \ IO =
        region _ {
            new MutList(Static): MutList[String, Static]
        }
```

should become something like:

```
    @Test
    def testNewMutList1Region03(): Unit =
        region r {
            discard new MutList(r): MutList[String, _]
        }
```

**Comment by paulbutcher** [comment]
Question about regions. I'm looking at [`File.readLinesIter`](https://github.com/flix/flix/blob/master/main/src/library/File.flix#L414-L448). I've managed to get it working as follows:

```
    pub def readLinesIter(r: Region[r], f: String): Iterator[Result[String, String], r] \ { Read(r), Write(r), IO } =
        try {
            import new java.io.File(String): ##java.io.File \ IO as newFile;
            import java.io.File.toPath(): ##java.nio.file.Path \ IO;
            import static java.nio.file.Files.newBufferedReader(##java.nio.file.Path): ##java.io.BufferedReader \ IO;
            import java.io.BufferedReader.readLine(): String \ Read(r);

            let javaFile = newFile(f);
            let javaPath = toPath(javaFile);
            let reader = newBufferedReader(javaPath);
            let line = ref readLine(reader) @ r;

            let done = () -> Object.isNull(deref line);
            let next = () -> {
                try {
                    let l = deref line;
                    line := readLine(reader);
                    Ok(l)
                } catch {
                    case ex: ##java.io.IOException => {
                        line := unsafe_cast null as String;
                        import java.lang.Throwable.getMessage(): String \ {};
                        Err(getMessage(ex))
                    }
                }
            };
            Iterator(done, next)
        } catch {
            case ex: ##java.io.IOException =>
                import java.lang.Throwable.getMessage(): String \ IO;
                Iterator.singleton(r, Err(getMessage(ex)))
        }
```

But I'm not convinced that this isn't abusing regions. I'm importing `readLine` as follows:

```
import java.io.BufferedReader.readLine(): String \ Read(r);
```

Which feels like a lie: I feel like it should really be:

```
import java.io.BufferedReader.readLine(): String \ IO;
```

But if I do that then I get:

```
-- Type Error -------------------------------------------------- File.flix

>> Mismatched Pure and Effectful Functions.

443 |             Iterator(done, next)
                  ^^^^^^^^^^^^^^^^^^^^
                  mismatched pure and effectful functions.

Type One: ((Done[b1], Next[t0, b1])) -> Iterator[t0, b1]
Type Two: ((Unit -> Bool & r, Unit -> Result[String, String] & Impure)) -> t0
```

What's the right approach here?

**Comment by paulbutcher** [comment]
Perhaps this (still lying) is closer:

```
    pub def readLinesIter(r: Region[r], f: String): Iterator[Result[String, String], r] \ { Read(r), Write(r), IO } =
        try {
            import new java.io.File(String): ##java.io.File \ IO as newFile;
            import java.io.File.toPath(): ##java.nio.file.Path \ IO;
            import static java.nio.file.Files.newBufferedReader(##java.nio.file.Path): ##java.io.BufferedReader \ IO;
            import java.io.BufferedReader.readLine(): String \ IO;

            let javaFile = newFile(f);
            let javaPath = toPath(javaFile);
            let reader = newBufferedReader(javaPath);
            let line = ref readLine(reader) @ r;

            let done = () -> Object.isNull(deref line);
            let next = () -> {
                try {
                    let l = deref line;
                    line := unsafe_cast readLine(reader) as _ \ Read(r);
                    Ok(l)
                } catch {
                    case ex: ##java.io.IOException => {
                        line := unsafe_cast null as String;
                        import java.lang.Throwable.getMessage(): String \ {};
                        Err(getMessage(ex))
                    }
                }
            };
            Iterator(done, next)
        } catch {
            case ex: ##java.io.IOException =>
                import java.lang.Throwable.getMessage(): String \ {};
                Iterator.singleton(r, Err(getMessage(ex)))
        }
```

**Comment by magnus-madsen** [comment]
You are correct (2x).

The issue is that `Iterator` is only using one of the two effect systems. 

Its okay to continue with the "lie" -- for now, we will clean it up :)

**Comment by paulbutcher** [comment]
The next wrinkle I need help with:

Given an interface with a method that takes an array parameter, we can implement it in the `Static` region:

```
    import static dev.flix.test.TestBoolArrayInterface.runTest(TestBoolArrayInterface): Bool;
    let anon = new TestBoolArrayInterface {
      def testMethod(_this: TestBoolArrayInterface, xs: Array[Bool, Static]): Array[Bool, Static] = ???
    }
```

But if we change the `Static`s above to a region, it fails.

The reason for this is [this](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Safety.scala#L774-L794) code in Safety which checks the method signatures to ensure that we're implementing the methods contained within the interface.

The problem is that `getFlixType`, given a Java array, returns `Type.Apply` with `Type.False` (i.e. `Static`) for the region. I suspect that it should be doing something else, but I'm not sure what that might be?

Suggestions gratefully received.

**Comment by magnus-madsen** [comment]
I think the issue can be understood at two levels:

Short-term: How can we have interop with Flix Arrays and Java Arrays? I think some hack inside `getFlixType` should be sufficient. We probably need to special case on the Array type constructor and take it from there.

Longer-term: What does it actually mean to have Java Arrays given that we have regions in Flix arrays?
- Do we want two separate types? (with means to convert between them)
- Do we want to rely less on Array and use a new type (forthcoming Vector) to pass data back and forth?

**Comment by magnus-madsen** [comment]
How many uses of `Scoped` remain?

**Comment by paulbutcher** [comment]
> How many uses of `Scoped` remain?

It's basically just `Iterable.iterator` now (which will require changes to the code generation as discussed).

**Comment by magnus-madsen** [comment]
> > How many uses of `Scoped` remain?
> 
> It's basically just `Iterable.iterator` now (which will require changes to the code generation as discussed).

What changes? My mind is completely blanking :)

**Comment by paulbutcher** [comment]
> What changes? My mind is completely blanking :)

https://github.com/flix/flix/issues/5131#issuecomment-1361271948

**Comment by paulbutcher** [comment]
I've got these changes on my todo list, but I'm currently looking at getting rid of `Static` use within the Channel code, which also requires code changes (and is proving more tricky than I hoped).

**Comment by magnus-madsen** [comment]
Ah okay- Its not to the code generated, its to the desugaring of `foreach`. I don't think that should be too difficult. 

We essentially need to implement the part where we wrap the foreach in a new LetRegion.

**Comment by paulbutcher** [comment]
Yup, I'm not concerned I won't be able to do them, just haven't got there yet.

**Comment by paulbutcher** [comment]
> Short-term: How can we have interop with Flix Arrays and Java Arrays? I think some hack inside getFlixType should be sufficient. We probably need to special case on the Array type constructor and take it from there.

Hmm. Thinking about it, we could handle this within Safety. Right now we're doing a simple equality check on the parameter and return types, and we probably want to do something closer to checking that the structure of the types is equivalent.

**Comment by paulbutcher** [comment]
For reference, the last remaining uses of `Static` are:

* In the tests for Java Functions (requires additional work to remove as discussed in #5172)
* In the channel code (requires additional work to remove as discussed in #5177)
* In the Fixpoint code (not sure how difficult this will be to remove)
* A few straggling tests (most of which can probably just be deleted?)

**Comment by paulbutcher** [comment]
Before I push ahead with the `Iterable` changes, do I understand correctly that this is the signature we want for `Iterable`?

(although obviously this is temporary pending #5132)

```
///
/// A type class for mutable data structures that can be iterated.
///
pub class Iterable[t: Type -> Region -> Type] {

    ///
    /// Returns an iterator over `t` in region `r1`.
    ///
    pub def iterator(r1: Region[r1], t: t[a, r2]): Iterator[a, r1] \ { Write(r1), Read(r2) }

    ///
    /// Returns an iterator over `t` in region `r1` zipped with the indices of the elements.
    ///
    pub def enumerator(r1: Region[r1], t: t[a, r2]): Iterator[(a, Int32), r1] \ { Write(r1), Read(r2) } =
        Iterable.iterator(t) |> Iterator.zipWithIndex

}
```

**Comment by paulbutcher** [comment]
Hmm, or perhaps the signature of `iterator` should be:

```
pub def iterator(r1: Region[r1], t: t[a, r2]): Iterator[a, r1 and r2] \ { Write(r1), Read(r2) }
```

?

**Comment by magnus-madsen** [comment]
> Hmm, or perhaps the signature of `iterator` should be:
> 
> ```
> pub def iterator(r1: Region[r1], t: t[a, r2]): Iterator[a, r1 and r2] \ { Write(r1), Read(r2) }
> ```
> 
> ?

This is correct. I think we should drop the read/write part for now. Point is that it touches r1 and r2.

**Comment by paulbutcher** [comment]
@magnus-madsen I'm looking at the remaining uses of `Scoped` and `regionOf`. Most of them are in `Iterator.flix`.

But if I'm understanding correctly, it looks to me as though there's more that needs changing here than just those functions that use `Scoped`? 

So, for example, `zipWithIndex` *does* use `Scoped` and should (I believe) be changed from its current implementation:

```
    @Lazy
    pub def zipWithIndex(iter: Iterator[a, r]): Iterator[(Int32, a), r] \ Write(r) =
        let Iterator(done, next) = iter;
        let i = ref 0 @ Scoped.regionOf(iter);
        let next1 = () -> {
            let x = (deref i, next());
            i := deref i + 1;
            x
        };
        Iterator(done, next1)
```

to take a region parameter as follows:

```
pub def zipWithIndex(r1: Region[r1], iter: Iterator[a, r2]): Iterator[(Int32, a), r1 and r2] \ Write(r2) = ???
```

But then there are lots of other functions which don't use `Scoped` such as `append`:

```
    @Lazy
    pub def append(iterA: Iterator[a, r1], iterB: Iterator[a, r2]): Iterator[a, r1 and r2] =
        let Iterator(doneA, nextA) = iterA;
        let Iterator(doneB, nextB) = iterB;
        let doneAB = () -> doneA() and doneB();
        let nextAB = () -> if (doneA()) nextB() else nextA();
        Iterator(doneAB, nextAB)
```

Which, although it doesn't use `Scoped` does return a new `Iterator`. So I think that these functions should also be changed along similar lines? So `append` would become:

```
pub def append(r3: Region[r3], iterA: Iterator[a, r1], iterB: Iterator[a, r2]): Iterator[a, r1 and r2 and r3] = ???
```

Do I understand that correctly?

**Comment by magnus-madsen** [comment]
I think you have identified an interesting philosophical issue. 

For `Iterator.append`, how many regions should we have? We have lots of options:

- We could require that both iterators live in the same region and we could place the resulting iterator in the same region. This may seem overly restrictive, but at least it avoids the need to pass a region to `Iterator.append`.
- We could store each region with the iterator (like e.g. MutDeque) and then we could reuse the region in `Iterator.append` (e.g. reusing either the first or second).
- We could take a new region argument, thus three regions would be in play.

TLDR:

```
def append(iter1: Iterator[t, r], iter2: Iterator[t, r]): Iterator[t, r]
def append(iter1: Iterator[t, r1], iter2: Iterator[t, r2]): Iterator[t, r1 and r2]
def append(rh: Region[r], iter1: Iterator[t, r1], iter2: Iterator[t, r2]): Iterator[t, r and r1 and r2]
```

**Comment by paulbutcher** [comment]
> * We could require that both iterators live in the same region and we could place the resulting iterator in the same region. This may seem overly restrictive, but at least it avoids the need to pass a region to Iterator.append.

This definitely seems excessively restrictive to me.

I would be OK with either of the other two options. How do we choose?

**Comment by magnus-madsen** [comment]
The new `Iterator` is in, so maybe now we can get rid of `Scoped`?

**Comment by paulbutcher** [comment]
I will take a look 👍
